fairscale/experimental/nn/mevo.py

```diff
@@ -0,0 +1,483 @@
+            print("DEBUG target bwd")
+        assert len(args) == 1
+        i, w, target = ctx.saved_tensors
+        assert i.requires_grad
+        assert w.requires_grad
+        assert not target.requires_grad
+        i = i.detach().requires_grad_(True)
+        w = w.detach().requires_grad_(True)
+        with torch.enable_grad():
+            scores = TargetScoreFunction.get_target_score(i, w, target, ctx.kernel_obj.fp_target)
+        torch.autograd.backward(scores, *args)
+        if ctx.kernel_obj.proj_weight.grad is not None:
+            # This means we accumulate full grad between iters. Not memory efficient.
+            ctx.kernel_obj.proj_weight.grad.add_(w.grad)
+        else:
+            ctx.kernel_obj.proj_weight.grad = w.grad
+        return i.grad, None, None, None
+
+
+class BackwardTriggerFn(torch.autograd.Function):
+    """A backward trigger function."""
+
+    @staticmethod
+    def forward(  # type: ignore
+        ctx: Any, w: torch.Tensor, trigger_tensor: torch.Tensor
+    ) -> torch.Tensor:
+        """We take a weight tensor and the trigger as inputs and output the weight directly."""
+        if DEBUG and dist.is_initialized() and dist.get_rank() == 0:
+            print("DEBUG trigger fwd")
+        ctx.save_for_backward(w, trigger_tensor)
+        return w
+
+    @staticmethod
+    def backward(ctx: Any, *args: Any) -> Any:
+        """We return zero grad for the trigger only."""
+        if DEBUG and dist.is_initialized() and dist.get_rank() == 0:
+            print("DEBUG trigger bwd")
+        assert len(args) == 1
+        w, trigger = ctx.saved_tensors
+        assert w.requires_grad
+        assert trigger.requires_grad
+        return None, torch.zeros_like(trigger)
+
+
+class BackwardTrigger(nn.Module):
+    """A backward trigger module.
+
+       This module takes a parameter as an input and create a linked parameter
+       from a newly created trigger parameter.
+
+       The way to use it in a module's ``__init__'' and ``forward'' functions:
+
+       ```
+       def __init__():
+         ...
+         self.trigger = BackwardTrigger(some_layer.weight)
+         ...
+
+       def forward():
+         w = self.trigger()
+         ... continue to use w ...
+       ```
+
+       As a resule, the trigger's backward hook will be called at the end of
+       the backward for the module that uses this trigger.
+    """
+
+    def __init__(self, linked_param: torch.Tensor):
+        super().__init__()
+        assert isinstance(linked_param, nn.Parameter)
+        self.trigger = nn.Parameter(torch.rand(1, dtype=linked_param.dtype))
+        self.trigger._linked_param = linked_param
+
+    def forward(self) -> torch.Tensor:  # type: ignore
+        return BackwardTriggerFn.apply(self.trigger._linked_param, self.trigger)
+
+
+class MemoryEfficientVocabOutput(nn.Module):  # AKA. MEVO
+    """ Fused fc + softmax + nll_loss in a tiled fashion.
+
+        This uses much less memory but is quite a bit slower.
```

Comment:
just curious about why it is slower. do we know why?

Reply:
Yes, we know some of it is due to re-computation. Others are likely due to python looping. This is definitely an important future work.